### PR TITLE
try_cast implementation

### PR DESCRIPTION
**Pull request by kovrus**

### CODE REVIEW

File: sql-parser/src/main/java/io/crate/sql/tree/TryCast.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+ * agreement.
+ */
+
+package io.crate.sql.tree;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class TryCast
+        extends Expression
+{
```
Comment (dobe) [inline sql-parser/src/main/java/io/crate/sql/tree/TryCast.java]: pls check your formatting settings the { should be kept on the initial line

File: sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java [hunk 2]
```diff
@@ -431,13 +431,28 @@ protected Symbol visitCast(Cast node, ExpressionAnalysisContext context) {
...
+            FunctionInfo functionInfo = CastFunctionResolver.functionInfo(symbolToCast.valueType(), returnType);
+
+            try {
+                if (symbolToCast.symbolType().isValueSymbol()) {
+                    return Literal.convert(symbolToCast, returnType);
+                }
+                return context.allocateFunction(functionInfo, Arrays.asList(symbolToCast));
+            } catch (ClassCastException | IllegalArgumentException e) {
+                return Literal.NULL;
+            }
```
Comment (dobe) [inline sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: cast will fail on real data rows, should return null

File: sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java [hunk 3]
```diff
@@ -431,13 +432,34 @@ protected Symbol visitCast(Cast node, ExpressionAnalysisContext context) {
         }
 
         @Override
+        protected Symbol visitTryCast(TryCast node, ExpressionAnalysisContext context) {
+            DataType returnType = DATA_TYPE_ANALYZER.process(node.getType(), null);
+            Symbol sourceSymbol = process(node.getExpression(), context);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: isn't it possible to call `cast(sourceSymbol, targetType, context)` here and just wrap that + catch exception?

File: sql/src/test/java/io/crate/integrationtests/CastIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+import static org.hamcrest.collection.IsArrayContainingInOrder.arrayContaining;
+
+
+public class CastIntegrationTest extends SQLTransportIntegrationTest {
+
+    @Rule
+    public ExpectedException expectedException = ExpectedException.none();
+
+    @Test
+    public void testTryCastValidLiteralCasting() {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/CastIntegrationTest.java]: why do we need so many integration tests?  `cast(..)` is already tested ... imho 1 integration tests is enough. The rest can be done with unit tests

File: sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java [hunk 5]
```diff
@@ -433,13 +434,31 @@ protected Symbol visitCast(Cast node, ExpressionAnalysisContext context) {
         }
 
         @Override
+        protected Symbol visitTryCast(TryCast node, ExpressionAnalysisContext context) {
+            DataType returnType = DATA_TYPE_ANALYZER.process(node.getType(), null);
+            if (!CastFunctionResolver.supportsExplicitConversion(returnType)) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: why do you negate it?

File: sql/src/main/java/io/crate/operation/scalar/cast/CastFunctionResolver.java [hunk 6]
```diff
@@ -67,4 +67,8 @@ public static FunctionInfo functionInfo(DataType dataType, DataType returnType)
         return new FunctionInfo(new FunctionIdent(functionName, ImmutableList.of(dataType)), returnType);
     }
 
+    public static boolean supportsExplicitConversion(DataType returnType) {
+        return functionMap.get(returnType) == null;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/scalar/cast/CastFunctionResolver.java]: method name implies that it returns true if it DOES support it ?

**Comment by mfussenegger** [comment]
otherwise ack

File: sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java (hunk 5 above)
Comment (kovrus) [inline sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: my mistake i did that ``functionMap.get(returnType) != null`` in supportsExplicitConversion method

**Pull request merged**